src/components/TodoItem.js

```diff
@@ -1,23 +1,67 @@
 import React from "react";
 import "./TodoItem.css";
-class TodoItem extends React.Component {
-  onItemChecked = e => {
-    this.props.onItemChecked(this.props.item.id);
+import Modal from "./Modal";
+import TextInput from "./TextInput";
+
+class TodoItem extends React.PureComponent {
+  state = { editing: false };
+  onChecked = e => {
+    this.props.onChecked(this.props.item.id);
+  };
+
+  onClick = e => {
+    this.props.onDelete(this.props.item.id);
+  };
+
+  onEdit = () => {
+    this.setState({ editing: true });
   };
-  onButtonClick = e => {
-    this.props.onItemDelete(this.props.item.id);
+
+  onClose = () => {
+    this.setState({ editing: false });
+  };
+
+  onChange = e => {
+    this.props.onCaptionChange(this.props.item.id, e.target.value);
   };
+
   render() {
-    const { item } = this.props;
     return (
       <div className="todo-item">
-        <input
-          onChange={this.onItemChecked}
-          type="checkbox"
-          checked={item.completed}
-        />
-        <div>{item.name}</div>
-        <button onClick={this.onButtonClick}>X</button>
+        {this.state.editing ? (
+          <Modal onClose={this.onClose}>
+            <TextInput
+              text={this.props.item.caption}
+              onChange={this.onChange}
+            />
+
+            <div>notifications</div>
+            <div>timer</div>
+          </Modal>
+        ) : null}
```

Comment:
Try to move the logic for rendering modal to `TodoList` component, since there can be only one active modal at a moment.